Source/Core/VideoBackends/D3D12/D3DUtil.cpp

```diff
@@ -0,0 +1,895 @@
+	"};\n"
+	"float4 main( PS_INPUT input ) : SV_Target\n"
+	"{\n"
+	"	return tex2D.Sample( linearSampler, input.tex ) * input.col;\n"
+	"};\n"
+};
+
+constexpr const char fontvertshader[] = {
+	"struct VS_INPUT\n"
+	"{\n"
+	"	float4 pos : POSITION;\n"
+	"	float4 col : COLOR;\n"
+	"	float2 tex : TEXCOORD;\n"
+	"};\n"
+	"struct PS_INPUT\n"
+	"{\n"
+	"	float4 pos : SV_POSITION;\n"
+	"	float4 col : COLOR;\n"
+	"	float2 tex : TEXCOORD;\n"
+	"};\n"
+	"PS_INPUT main( VS_INPUT input )\n"
+	"{\n"
+	"	PS_INPUT output;\n"
+	"	output.pos = input.pos;\n"
+	"	output.col = input.col;\n"
+	"	output.tex = input.tex;\n"
+	"	return output;\n"
+	"};\n"
+};
+
+int CD3DFont::Init()
+{
+	// Create vertex buffer for the letters
+
+	// Prepare to create a bitmap
+	unsigned int* pBitmapBits;
+	BITMAPINFO bmi;
+	ZeroMemory(&bmi.bmiHeader, sizeof(BITMAPINFOHEADER));
+	bmi.bmiHeader.biSize        = sizeof(BITMAPINFOHEADER);
+	bmi.bmiHeader.biWidth       =  (int)m_tex_width;
+	bmi.bmiHeader.biHeight      = -(int)m_tex_height;
+	bmi.bmiHeader.biPlanes      = 1;
+	bmi.bmiHeader.biCompression = BI_RGB;
+	bmi.bmiHeader.biBitCount    = 32;
+
+	// Create a DC and a bitmap for the font
+	HDC hDC = CreateCompatibleDC(nullptr);
+	HBITMAP hbmBitmap = CreateDIBSection(hDC, &bmi, DIB_RGB_COLORS, (void**)&pBitmapBits, nullptr, 0);
+	SetMapMode(hDC, MM_TEXT);
+
+	// create a GDI font
+	HFONT hFont = CreateFont(24, 0, 0, 0, FW_NORMAL, FALSE,
+							FALSE, FALSE, DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,
+							CLIP_DEFAULT_PRECIS, PROOF_QUALITY,
+							VARIABLE_PITCH, _T("Tahoma"));
+
+	if (nullptr == hFont)
+		return E_FAIL;
+
+	HGDIOBJ hOldbmBitmap = SelectObject(hDC, hbmBitmap);
+	HGDIOBJ hOldFont = SelectObject(hDC, hFont);
+
+	// Set text properties
+	SetTextColor(hDC, 0xFFFFFF);
+	SetBkColor  (hDC, 0);
+	SetTextAlign(hDC, TA_TOP);
+
+	TEXTMETRICW tm;
+	GetTextMetricsW(hDC, &tm);
+	m_line_height = tm.tmHeight;
+
+	// Loop through all printable characters and output them to the bitmap
+	// Meanwhile, keep track of the corresponding tex coords for each character.
+	int x = 0, y = 0;
+	char str[2] = "\0";
+	for (int c = 0; c < 127 - 32; c++)
+	{
+		str[0] = c + 32;
+		SIZE size;
+		GetTextExtentPoint32A(hDC, str, 1, &size);
+		if ((int)(x + size.cx + 1) > m_tex_width)
+		{
+			x  = 0;
+			y += m_line_height;
+		}
+
+		ExtTextOutA(hDC, x+1, y+0, ETO_OPAQUE | ETO_CLIPPED, nullptr, str, 1, nullptr);
+		m_tex_coords[c][0] = ((float)(x + 0))/m_tex_width;
+		m_tex_coords[c][1] = ((float)(y + 0))/m_tex_height;
+		m_tex_coords[c][2] = ((float)(x + 0 + size.cx))/m_tex_width;
+		m_tex_coords[c][3] = ((float)(y + 0 + size.cy))/m_tex_height;
+
+		x += size.cx + 3;  // 3 to work around annoying ij conflict (part of the j ends up with the i)
+	}
+
+	// Create a new texture for the font
+	// possible optimization: store the converted data in a buffer and fill the texture on creation.
+	// That way, we can use a static texture
+	std::unique_ptr<byte> texInitialData(new byte[m_tex_width * m_tex_height * 4]);
```

Comment:
This is still technically undefined behavior. This unique_ptr will only call `delete`, not `delete[]`, this needs to either be a vector or a `unique_ptr<byte[]>`